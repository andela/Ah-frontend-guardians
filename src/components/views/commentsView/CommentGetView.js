import React, { Component } from 'react'
import { connect } from 'react-redux'
import CommentReplyForm from "../../presentational/coments/CommentReplyForm";
import { getcommentAction } from '../../../action/actionCreator/comments/getCommentAction'
import {updateCommentAction} from '../../../action/actionCreator/comments/updateComment'
import {deleteCommentAction} from '../../../action/actionCreator/comments/deleteCommentAction';

import swal from "sweetalert";


export class GetComments extends Component {
  constructor(props) {
      super(props)
      this.state = {
        commentstate: '',
        errors: '',
        data: '',
        slug: '',
      };
  }

  componentDidMount() {
    const slug = this.props.slug;
    this.props.getcommentAction(slug);
  }

  componentWillReceiveProps(nextProps) {
    const {updateComment, deleteComment} = nextProps
    if (updateComment){
      this.setState({updateComment:updateComment})
    }
    if (deleteComment){
      this.setState({deleteComment:deleteComment})
    }
  }

  onChange(e) {
    this.setState({
      commentstate: e.target.value
    })
  }

  onSubmit=(e) => {
    e.preventDefault() 
    const data = {
      body: this.state.commentstate,
     }
  const slug = this.props.slug;
  const commentId = e.target.id
  this.props.updateCommentAction(slug, data, commentId);
  }

  handleDelete = (e) => {
  const slug = this.props.slug;
  const commentId = e.target.id
  swal({
    title: "Are you sure?",
    text: "Once deleted, you will not be to recover this comment!",
    icon: "warning",
    buttons: true,
    dangerMode: true
  }).then((willDelete) => {
    if (willDelete) {
      this.props.deleteCommentAction(slug, commentId)
    }
  });
  }

  render() {
const username= window.localStorage.getItem('username') 

const comments = this.props.comments.map((comment) => (
  <div key={comment.id} id = "getCommentdiv">
    <p id = "getCommentId">
      <i className="fas fa-user fa-2x icon_color"/>
      <span id="spanId">
        {comment.author.username}
      </span><br/>
      <label>
        <font color="black">{comment.body}</font>
      </label>
    </p>
    {(comment.author.username == username) ?
    <div className="editDiv">
      <a className="editLink" data-toggle="collapse" href={"#collapseExample3"+comment.id} role="button" aria-expanded="false" aria-controls="collapseExample2">
        Edit
      </a>
      <a href="#" onClick={this.handleDelete} id={comment.id} className="deleteLink">
        Delete
      </a>
    </div> : ''}
    <div className="collapse" id={'collapseExample3'+comment.id} >
      <form onSubmit={this.onSubmit.bind(this)} id={comment.id}>
        <textarea className="editComment" id={comment.id} name = "comment" onChange={this.onChange.bind(this)} defaultValue={comment.body}/>
        <input type = "submit"className="comment_update_button" value="Save" />
      </form>
    </div>
  </div>
  ))
      
return (
      <div >
        <div>{comments.length < 1 ? <p>No comments</p>:comments}</div>
            <CommentReplyForm
            />
      </div>
    )
  }
}

const mapStatetoProps = (state) => ({
    comments: state.getarticlecomments.data,
    updateComment: state.updatecomments,
    deleteComment: state.deleteCommentAction,
   })

export default connect(mapStatetoProps, { getcommentAction, updateCommentAction, deleteCommentAction})(GetComments)
