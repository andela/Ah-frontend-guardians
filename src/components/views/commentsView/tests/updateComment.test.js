import React from 'react';
import { shallow } from 'enzyme';
import { GetComments , mapStateToProps} from "../CommentGetView";

const props={
  getcommentAction:jest.fn(),
  updateCommentAction:jest.fn(),
  deleteCommentAction:jest.fn(),
  updateComment:"UPDATED",
  deleteComment:"DELETED",
  slug:"test",
  comments:[{id:4 , body:"this is an article", author: { username:"guardians"}, }],
}

const e ={
  preventDefault:()=>{},
  target: { id: 1, value: 123 }
}
localStorage.setItem('token','token')
let wrapper 
  describe(`CommentView`, () => {

    it("Comment form should render handleFormSubmit", () => {  
     wrapper = shallow(<GetComments {...props}/>)
      wrapper.instance().onSubmit(e)
      wrapper.instance().handleDelete(e)
      expect(props.updateCommentAction).toBeCalled()
    })

    it("dcbdg", () => {  
        wrapper = shallow(<GetComments {...props}/>)
        wrapper.instance().onChange(e)
        expect(wrapper.instance().state.commentstate).toEqual(123)
        })
       
    
    it("Comment form should render", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("shdjbsad", () =>{
        wrapper = shallow(<GetComments {...props}/>)
        wrapper.setProps(props)
        expect(wrapper.instance().state.updateComment).toEqual("UPDATED")
        expect(wrapper.instance().state.deleteComment).toEqual("DELETED")
    });

    describe('When the form is submitted', () => {
        // it('should call the mock comment function', () => {
        //     wrapper.find('form').simulate('submit',
        //         { preventDefault() { } })                
        //     expect(props.postcommentAction.mock.calls.length).toBe(1);
        // });

//         it('should be called with the comment in the state as arguments', () => {
//             const simulator = (name, value) => wrapper.find('CommentForm').dive().find('#comment').simulate(
//                     'change', {
//                         target:
//                             { name, value },
//                     },
//                 )

//             simulator('comment', 'Good article')
//             wrapper.find('CommentForm').dive().find('#comment_form_container').simulate(
//                 'submit',
//                 { preventDefault() { } },
//             )
//             expect(props.postcommentAction.mock.calls[0][1]).toEqual(
//                 {
//                     body: 'Good article',
//                 },
//             )
//         })
//         it('should will recieve error comment props', () => {
//             const newState = { errors: "efgrt" }

//               wrapper.setProps(newState)
//             console.log(wrapper.state());

//               expect(wrapper.state()).toEqual({ "body": "", "errors": "efgrt", "commentstate": '', "data":"" })
//           })
//           it('should will recieve data comment props', () => {
//             const newState = { data: "efgrt", errors:"" }

//               wrapper.setProps(newState)
//               expect(wrapper.state()).toEqual({ "errors": "", "data": "efgrt", "commentstate": '', "body":""})
//           })
//           it('handles the state changes on the comment Form', () => {
//             const commentField = wrapper.find('CommentForm').dive().find('#comment_form_container')

//             commentField.value = 'Looks good to me';
//             const event = { target: { name: 'comment', value: 'Looks good to me' } };

//             commentField.simulate('change', event);
//             expect(commentField.value).toEqual('Looks good to me');
//           });
 })
})