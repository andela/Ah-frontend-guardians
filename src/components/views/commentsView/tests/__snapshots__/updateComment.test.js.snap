// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentView Comment form should render 1`] = `ShallowWrapper {}`;
